File: app/src/main/java/com/example/my_board/ListView/ExListViewAdapter.java
package com.example.my_board.ListView;

import android.content.Context;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseExpandableListAdapter;
import android.widget.Button;
import android.widget.ExpandableListView;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.example.my_board.Holder;
import com.example.my_board.ListViewItem;
import com.example.my_board.R;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.util.ArrayList;
import java.util.HashMap;

public class ExListViewAdapter<ChildListViewHolder> extends BaseExpandableListAdapter {
    private ImageView board_icon;
    private TextView board_title;
    private TextView board_uid;
    private String uid;
    private String title;
    private ExpandableListView listView;
    private LinearLayout likePart;


    private ArrayList<ListViewItem> mParentList;
    private HashMap<ListViewItem, ArrayList<ListViewItem>> mChildHashMap;

    //ListViewAdapter의 생성자
    public ExListViewAdapter(ArrayList<ListViewItem> parentList, HashMap<ListViewItem, ArrayList<ListViewItem>> childHashMap, ExpandableListView listView){
        this.mParentList = parentList;
        this.mChildHashMap = childHashMap;
        this.listView = listView;
    }


    public void setmParentList(ArrayList<ListViewItem> parentList){
        this.mParentList = parentList;
    }

    public void setmChildHashMap(HashMap<ListViewItem, ArrayList<ListViewItem>> childHashMap){
        this.mChildHashMap = childHashMap;
    }



    public void setUId(String uid){
        this.uid = uid;
    }

    public void setBoard_title(String title){
        this.title = title;
    }


    @Override
    public int getGroupCount() {
        return mParentList.size();
    }

    @Override
    public int getChildrenCount(int groupPosition) {
        if(this.mChildHashMap.get(this.mParentList.get(groupPosition)) != null){
            return this.mChildHashMap.get(this.mParentList.get(groupPosition)).size();
        }
        else{
            return 0;
        }
    }

    @Override
    public ListViewItem getGroup(int groupPosition) {
        return mParentList.get(groupPosition);
    }

    @Override
    public ListViewItem getChild(int groupPosition, int childPosition) {
        return this.mChildHashMap.get(this.mParentList.get(groupPosition)).get(childPosition);
    }

    @Override
    public long getGroupId(int groupPosition) {
        return groupPosition;
    }

    @Override
    public long getChildId(int groupPosition, int childPosition) {
        return childPosition;
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

    @Override
    public View getGroupView(final int groupPosition, boolean isExpanded, View convertView, final ViewGroup parent) {
        final ListViewItem listViewItem = mParentList.get(groupPosition);
        System.out.println("로그인한 회원의 id는 " + uid + ", 게시판을 작성한 id는 " + listViewItem.getBoard_uid());
        Holder holder;
        System.out.println("가져온 부모 listView는 " +  listViewItem.getBoard_title());
        if(convertView == null){
            Context context = parent.getContext();
            holder = new Holder();
            holder.boardTitle = listViewItem.getBoard_title();
            LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            convertView = inflater.inflate(R.layout.listview_mcitem, parent, false);
            holder.deleteBtn = (Button) convertView.findViewById(R.id.content_delete);
            if(!listViewItem.getBoard_uid().equals(uid)) {
                holder.deleteBtn.setVisibility(View.INVISIBLE);
            }
            holder.editBtn = (Button) convertView.findViewById(R.id.cComent);
            holder.editText = (TextInputEditText) convertView.findViewById(R.id.TextInputEditText_cComment);
            convertView.setTag(holder);
        }else{
            holder = (Holder) convertView.getTag();
            if(!listViewItem.getBoard_uid().equals(uid)) {
                holder.deleteBtn.setVisibility(View.INVISIBLE);
            }else{
                holder.deleteBtn.setVisibility(View.VISIBLE);
            }
        }
        final Holder finalHolder = holder;
        holder.position = groupPosition;


        holder.editText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                Log.d("TAG", "새로운 값이 업데이트되었다.");
                mParentList.get(finalHolder.position).setBoard_cComment(s.toString());
            }
        });
        if(listViewItem.getBoard_uid().equals(uid)) {
            holder.deleteBtn.setOnClickListener(new Button.OnClickListener() {
                @Override
                public void onClick(View v) {
                    System.out.println("삭제가 된 것은 " +  mParentList.get(finalHolder.position).getBoard_title());
                    FirebaseDatabase database = FirebaseDatabase.getInstance();
                    DatabaseReference removeContent = database.getReference("Content/" + title + "/comment/" + uid +  mParentList.get(finalHolder.position).getBoard_title() + "/");
                    removeContent.removeValue();
                }
            });
        }
        holder.editBtn.setOnClickListener(new Button.OnClickListener(){
            @Override
            public void onClick(View v) {
                System.out.println("클릭된 것의 제목은 " + mParentList.get(finalHolder.position).getBoard_title());
                String content = mParentList.get(finalHolder.position).getBoard_cComment();
                System.out.println("content 내용은 " + content);
                if(content == null || content.isEmpty()){
                    finalHolder.editText.setText("please,write!");
                }else{
                    System.out.println("대댓글은 " + content + ", 댓글은 " +  mParentList.get(finalHolder.position).getBoard_title());
                    HashMap<Object, String> hashMap = new HashMap<>();
                    hashMap.put("comment", content);
                    hashMap.put("uid", uid);
                    FirebaseDatabase database = FirebaseDatabase.getInstance();
                    DatabaseReference reference = database.getReference("Content/" + title + "/comment/" + uid +  mParentList.get(finalHolder.position).getBoard_title() + "/");
                    reference.child("ccomment/" + uid + content + '/').setValue(hashMap);
                }
            }
        });


        //화면에 표시될 View(Layout이 infate된)으로부터 위젯에 대한 참조 획득
        board_icon = (ImageView) convertView.findViewById(R.id.board_img);
        board_title = (TextView) convertView.findViewById(R.id.board_title);
        board_uid = (TextView)convertView.findViewById(R.id.board_uid);

        //아이템 내 각 위젯에 데이터 반영
        board_icon.setImageDrawable(listViewItem.getBoard_icon());
        board_title.setText(listViewItem.getBoard_title());
        board_uid.setText(listViewItem.getBoard_uid());

        return convertView;
    }

    @Override
    public View getChildView(int groupPosition, int childPosition, boolean isLastChild, View convertView, ViewGroup parent) {
        final ListViewItem listViewItem = getChild(groupPosition, childPosition);
        if(convertView == null){
            Context context = parent.getContext();
            LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            convertView = inflater.inflate(R.layout.listview_item, parent, false);
        }

        //화면에 표시될 View(Layout이 infate된)으로부터 위젯에 대한 참조 획득
        board_icon = (ImageView) convertView.findViewById(R.id.board_img);
        board_title = (TextView) convertView.findViewById(R.id.board_title);
        board_uid = (TextView)convertView.findViewById(R.id.board_uid);
        likePart = (LinearLayout)convertView.findViewById(R.id.likePart);
        likePart.setVisibility(View.INVISIBLE);


        //아이템 내 각 위젯에 데이터 반영
        board_icon.setImageDrawable(listViewItem.getBoard_icon());
        board_title.setText(listViewItem.getBoard_title());
        System.out.println("제목, 내용 확인해보기" + listViewItem.getBoard_title() + ", " + listViewItem.getBoard_uid());
        board_uid.setText(listViewItem.getBoard_uid());

        return convertView;
    }

    @Override
    public boolean isChildSelectable(int groupPosition, int childPosition) {
        return true;
    }

}

File: app/src/main/java/com/example/my_board/ListView/ExListViewAdapter.kt
package com.example.my_board.ListView

import android.content.Context
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.example.my_board.Holder
import com.example.my_board.R
import com.google.android.material.textfield.TextInputEditText
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import java.util.*

class ExListViewAdapter(private var mParentList: ArrayList<ListViewItem>, private var mChildHashMap: HashMap<ListViewItem, ArrayList<ListViewItem>?>, private val listView: ExpandableListView) : BaseExpandableListAdapter() {
    private var board_icon: ImageView? = null
    private var board_title: TextView? = null
    private var board_uid: TextView? = null
    private var uid: String? = null
    private var character: String? = null
    private var title: String? = null
    fun setmParentList(parentList: ArrayList<ListViewItem>) {
        mParentList = parentList
    }

    fun setmChildHashMap(childHashMap: HashMap<ListViewItem, ArrayList<ListViewItem>?>) {
        mChildHashMap = childHashMap
    }

    fun setUId(uid: String?) {
        this.uid = uid
    }

    fun setCharacter(character: String?){
        this.character = character
    }

    fun setBoard_title(title: String?) {
        this.title = title
    }

    override fun getGroupCount(): Int {
        return mParentList.size
    }

    override fun getChildrenCount(groupPosition: Int): Int {
        return if (mChildHashMap[mParentList[groupPosition]] != null) {
            mChildHashMap[mParentList[groupPosition]]!!.size
        } else {
            0
        }
    }

    override fun getGroup(groupPosition: Int): ListViewItem {
        return mParentList[groupPosition]
    }

    override fun getChild(groupPosition: Int, childPosition: Int): ListViewItem {
        return mChildHashMap[mParentList[groupPosition]]!![childPosition]
    }

    override fun getGroupId(groupPosition: Int): Long {
        return groupPosition.toLong()
    }

    override fun getChildId(groupPosition: Int, childPosition: Int): Long {
        return childPosition.toLong()
    }

    override fun hasStableIds(): Boolean {
        return true
    }

    override fun getGroupView(groupPosition: Int, isExpanded: Boolean, convertView: View?, parent: ViewGroup?): View {
        var convertView = convertView
        val listViewItem = mParentList[groupPosition]
        val holder: Holder
        if (convertView == null) {
            val context = parent?.context
            val database = FirebaseDatabase.getInstance()
            val inflater = context?.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            holder = Holder()
            holder.boardTitle = listViewItem.board_title
            convertView = inflater.inflate(R.layout.listview_mcitem, parent, false)
            holder.deleteBtn = convertView.findViewById<View>(R.id.content_delete) as Button
            if (listViewItem.board_uid != uid) {
                holder.deleteBtn!!.visibility = View.INVISIBLE
            }
            holder.editBtn = convertView.findViewById<View>(R.id.cComent) as Button
            holder.editText = convertView.findViewById<View>(R.id.TextInputEditText_cComment) as TextInputEditText
            convertView.tag = holder
        } else {
            holder = convertView.tag as Holder
            if (listViewItem.board_uid != uid) {
                holder.deleteBtn!!.visibility = View.INVISIBLE
            } else {
                holder.deleteBtn!!.visibility = View.VISIBLE
            }
        }
        holder.position = groupPosition
        holder.editText!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                Log.d("TAG", "새로운 값이 업데이트되었다.")
                mParentList[holder.position].board_cComment = s.toString()
            }
        })
        if (listViewItem.board_uid == uid) {
            holder.deleteBtn!!.setOnClickListener {
                val database = FirebaseDatabase.getInstance()
                val removeContent = database.getReference("Content/" + title + "/comment/" + uid + mParentList[holder.position].board_title + "/")
                removeContent.removeValue()
            }
        }
        holder.editBtn!!.setOnClickListener {
            println("클릭된 것의 제목은 " + mParentList[holder.position].board_title)
            val content = mParentList[holder.position].board_cComment
            println("content 내용은 $content")
            if (content == null || content.isEmpty()) {
                holder.editText!!.setText("please,write!")
            } else {
                println("대댓글은 " + content + ", 댓글은 " + mParentList[holder.position].board_title)
                val hashMap = HashMap<Any, String?>()
                hashMap["comment"] = content
                hashMap["uid"] = uid
                hashMap["character"]=character
                val database = FirebaseDatabase.getInstance()
                val reference = database.getReference("Content/" + title + "/comment/" + mParentList[holder.position].board_uid + mParentList[holder.position].board_title + "/")
                reference.child("ccomment/$uid$content/").setValue(hashMap)
            }
        }

        board_icon = convertView!!.findViewById<View>(R.id.board_img) as ImageView
        board_title = convertView!!.findViewById<View>(R.id.board_title) as TextView
        board_uid = convertView.findViewById<View>(R.id.board_uid) as TextView

        //아이템 내 각 위젯에 데이터 반영
        board_icon!!.setImageDrawable(listViewItem.character)
        board_title!!.text = listViewItem.board_title
        board_uid!!.text = listViewItem.board_uid
        return convertView
    }

    override fun getChildView(groupPosition: Int, childPosition: Int, isLastChild: Boolean, convertView: View?, parent: ViewGroup?): View {
        var convertView = convertView
        val listViewItem = getChild(groupPosition, childPosition)
        if (convertView == null) {
            val context = parent?.context
            val inflater = context?.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            convertView = inflater.inflate(R.layout.listview_citem, parent, false)
        }

        //화면에 표시될 View(Layout이 infate된)으로부터 위젯에 대한 참조 획득
        board_icon = convertView!!.findViewById<View>(R.id.board_img) as ImageView
        board_title = convertView!!.findViewById<View>(R.id.board_title) as TextView
        board_uid = convertView.findViewById<View>(R.id.board_uid) as TextView

        //아이템 내 각 위젯에 데이터 반영
        board_icon!!.setImageDrawable(listViewItem.character)
        board_title!!.text = listViewItem.board_title
        println("제목, 내용 확인해보기" + listViewItem.board_title + ", " + listViewItem.board_uid)
        board_uid!!.text = listViewItem.board_uid
        return convertView
    }

    override fun isChildSelectable(groupPosition: Int, childPosition: Int): Boolean {
        return true
    }
}